bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/TextLayout.java

```diff
@@ -399,26 +400,26 @@ void computeRuns (GC gc) {
 			if (length > 1) {
 				int stop = j + length - 1;
 				if (stop < tabsLength) {
-					run.width += tabs[stop] - tabs[j];
+					run.width += DPIUtil.autoScaleUp(getDevice(), tabs[stop], gc.data.shell) - DPIUtil.autoScaleUp(getDevice(), tabs[j], gc.data.shell);
 				} else {
 					if (j < tabsLength) {
-						run.width += tabs[tabsLength - 1] - tabs[j];
+						run.width += DPIUtil.autoScaleUp(getDevice(), tabs[tabsLength-1], gc.data.shell) - DPIUtil.autoScaleUp(getDevice(), tabs[j], gc.data.shell);
 						length -= (tabsLength - 1) - j;
 					}
-					int lastTabWidth = tabsLength > 1 ? tabs[tabsLength-1] - tabs[tabsLength-2] : tabs[0];
+					int lastTabWidth = tabsLength > 1 ? DPIUtil.autoScaleUp(getDevice(), tabs[tabsLength-1], gc.data.shell) - DPIUtil.autoScaleUp(getDevice(), tabs[tabsLength-2], gc.data.shell) : DPIUtil.autoScaleUp(getDevice(), tabs[0], gc.data.shell);
 					run.width += lastTabWidth * (length - 1);
 				}
 			}
 		}
-		if (wrapWidth != -1 && lineWidth + run.width > wrapWidth && !run.tab && !run.lineBreak) {
+		if (DPIUtil.autoScaleUp(getDevice(), wrapWidth, gc.data.shell) != -1 && lineWidth + run.width > DPIUtil.autoScaleUp(getDevice(), wrapWidth, gc.data.shell) && !run.tab && !run.lineBreak) {
```

Comment:
DPIUtil.autoScaleUp(getDevice(), wrapWidth, gc.data.shell) != -1 -> scaling up here doesn't make sense, -1 is always used as marker not unset values, scaling it can break this check